### PR TITLE
Serialise complex objects correctly

### PR DESCRIPTION
**Pull request by laurence79**
This is essentially a reimplementation of the ResourceSerializer that copes with a couple of edge cases that have tripped me up in my own use:
- Circular references
- Inclusion of deeply nested objects

Take a look at ResourceGraphSerializerTests to see tests that pass with the new serialiser. The new serialiser also passes all the existing tests (with a couple of exceptions where I have overridden the existing test with comments as to why).

The new serialiser approach differs from the original by first building a graph of objects to serialise using the IncludingContext, and then generates the response JSON from that.

This is anecdotal but I've also seen modest performance improvements from this approach.

By default the original serialiser is still used unless a new configuration option is set
`
config.ConfigureJsonApi(new JsonApiConfiguration
{
    UseGraphSerializer = true
});
`

I appreciate this is a large change, but hopefully it will be a benefit to the project 😊

### CODE REVIEW

**Comment by joukevandermaas** [comment]
This sounds great, I'll try to make some time to review this soon. Thanks so much!

**Comment by laurence79** [comment]
Thanks @joukevandermaas

**Comment by laurence79** [comment]
Fixes #141

**Comment by bjornharrtell** [comment]
Seems like nice work @laurence79, looking forward to testing it. I wonder if it also resolves https://github.com/joukevandermaas/saule/issues/99?

**Comment by laurence79** [comment]
Thanks @bjornharrtell, yes it seems like it might because only the attributes and relationships declared in the resource are read from the target objects.

File: Saule/Http/JsonApiConfiguration.cs [hunk 1]
```diff
@@ -22,6 +22,11 @@ public class JsonApiConfiguration
         /// <summary>
         /// Gets the expressions that are used to evaluate filter queries on a per-type basis.
         /// </summary>
-        public QueryFilterExpressionCollection QueryFilterExpressions { get; } = new QueryFilterExpressionCollection();
+        public QueryFilterExpressionCollection QueryFilterExpressions { get; } = new QueryFilterExpressionCollection();
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the new experimental graph serializer is used.
+        /// </summary>
+        public bool UseGraphSerializer { get; set; } = false;
```
Comment (joukevandermaas) [inline Saule/Http/JsonApiConfiguration.cs]: I wonder if we actually need this; I would prefer we just swap out the serializer (provided that all the tests pass with the new one).

I don't think anyone would want to use the old serializer with this new one in place anyway.

File: Saule/Http/AllowsIncludeAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        /// No related resources should be included.
+        /// </summary>
+        None
+    }
+
+    /// <summary>
+    /// Indicates that the target action supports the use of the include query param.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method)]
+    public sealed class AllowsIncludeAttribute : ActionFilterAttribute
```
Comment (joukevandermaas) [inline Saule/Http/AllowsIncludeAttribute.cs]: Moving this out of `AllowsQueryAttribute` seems unrelated to serialization. What was the reason for this change?

Since I already released 1.6, this would be a breaking change. I don't really mind (we can figure out a migration scenario or just release 2.0), but I'm not sure I understand the purpose of this change from the diff.

File: Saule/ObjectExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            }
+
+            var propertyInfo = obj.GetType().GetProperty(propertyName);
+            if (propertyInfo != null)
+            {
+                return propertyInfo.GetValue(obj);
+            }
+
+            var fieldInfo = obj.GetType().GetField(propertyName);
+            if (fieldInfo != null)
```
Comment (joukevandermaas) [inline Saule/ObjectExtensions.cs]: I'm not sure we should be reading fields. Unless this is actually used somewhere, I'd prefer we remove it.

File: Saule/Serialization/ResourceGraphNodeKey.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Saule.Serialization
+{
+    internal class ResourceGraphNodeKey : IEquatable<ResourceGraphNodeKey>
```
Comment (joukevandermaas) [inline Saule/Serialization/ResourceGraphNodeKey.cs]: I'm thinking this could be a `struct`?

File: Saule/Serialization/ResourceGraph.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            ResourceGraphPathSet includePaths,
+            int depth)
+        {
+            if (obj == null)
+            {
+                // end of the line - obj will be a leaf node in our graph
+                return;
+            }
+
+            if (obj is IEnumerable)
```
Comment (joukevandermaas) [inline Saule/Serialization/ResourceGraph.cs]: How does this work with `IQueryable`? Does it evaluate the query to build the graph? If so, wouldn't that mean multiple queries to fetch the relationships? Maybe we need some tests for this stuff, since it seems pretty easy to break with a simple change.

File: Tests/Serialization/ResourceGraphSerializerTests.cs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            var target = GetSerializer(person, new PersonWithDefaultIdResource(),
+                GetUri(id: "123"), DefaultPathBuilder, null, null);
+            var result = target.Serialize();
+            _output.WriteLine(result.ToString());
+
+            var relationships = result["data"]["relationships"];
+            var job = relationships["job"];
+            var friends = relationships["friends"];
+
+            Assert.Equal(job["data"].Type.ToString(), "Null");
```
Comment (joukevandermaas) [inline Tests/Serialization/ResourceGraphSerializerTests.cs]: I think you can just assert on the enum value, i.e.: 

```csharp
Assert.Equal(job["data"].Type, JTokenType.Null);
```

File: Tests/Serialization/ResourceGraphSerializerTests.cs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            Assert.NotNull(friends);
+        }
+
+        public override void HandlesNullValues()
+        {
+            var person = new Person(id: "45");
+            var target = GetSerializer(person, DefaultResource,
+                GetUri(id: "45"), DefaultPathBuilder, null, new IncludingContext { DisableDefaultIncluded = true });
+            /**
+             *  Changed from a null IncludingContext in the original version
```
Comment (joukevandermaas) [inline Tests/Serialization/ResourceGraphSerializerTests.cs]: Does this mean the `IncludingContext` is never `null` anymore (seems fine, just wondering)?

File: Tests/Serialization/ResourceSerializerTests.cs [hunk 8]
```diff
@@ -9,25 +9,39 @@
...
 using Saule.Queries.Including;
+using Saule.Queries.Pagination;
 
 namespace Tests.Serialization
 {
+    [Trait("Serializer", "ResourceSerializer")]
     public class ResourceSerializerTests
     {
-        private readonly ITestOutputHelper _output;
+        internal readonly ITestOutputHelper _output;
```
Comment (joukevandermaas) [inline Tests/Serialization/ResourceSerializerTests.cs]: Maybe you mean `protected`? In any case, if we just remove the old serializer, this should not be necessary anymore.

File: Saule/Http/JsonApiConfiguration.cs (hunk 1 above)
Comment (laurence79) [inline Saule/Http/JsonApiConfiguration.cs]: Sure, I've moved the new to be in place of the old and removed the config option.

File: Saule/Serialization/ResourceGraphNodeKey.cs (hunk 4 above)
Comment (laurence79) [inline Saule/Serialization/ResourceGraphNodeKey.cs]: Agreed, change made.

File: Tests/Serialization/ResourceGraphSerializerTests.cs (hunk 6 above)
Comment (laurence79) [inline Tests/Serialization/ResourceGraphSerializerTests.cs]: Of course, silly me 🙄

File: Tests/Serialization/ResourceSerializerTests.cs (hunk 8 above)
Comment (laurence79) [inline Tests/Serialization/ResourceSerializerTests.cs]: Old serialiser replaced.

**Comment by laurence79** [comment]
AppVeyor had a timeout on package restore. I couldn't see a way to easily retry the build, so I bumped the Newtonsoft.Json version to 1.0.2, hope thats ok 😊

File: Saule/ObjectExtensions.cs (hunk 3 above)
Comment (laurence79) [inline Saule/ObjectExtensions.cs]: I didn't originally have this, but some tests failed e.g. "Handles recursive properties on resource objects", and I figured the old serialiser would serialise fields.

File: Saule/Serialization/ResourceGraph.cs (hunk 5 above)
Comment (laurence79) [inline Saule/Serialization/ResourceGraph.cs]: Good point, I'll write some tests to see how this behaves.

File: Tests/Serialization/ResourceGraphSerializerTests.cs (hunk 7 above)
Comment (laurence79) [inline Tests/Serialization/ResourceGraphSerializerTests.cs]: No, I think the purpose of this test was to prove that if DisableDefaultIncluded was true then no relationships were returned.

It's entirely possible that I've misread this of course 😆

File: Saule/Http/AllowsIncludeAttribute.cs (hunk 2 above)
Comment (laurence79) [inline Saule/Http/AllowsIncludeAttribute.cs]: Good spot. This is not needed, I've reverted it to how it was before.

File: Saule/Serialization/ResourceGraph.cs (hunk 5 above)
Comment (laurence79) [inline Saule/Serialization/ResourceGraph.cs]: Ok perhaps I'm not completely clear on this. 

The expression tree is already evaluated and a response generated before the graph starts reflecting upon it. Is this what you meant @joukevandermaas?

Comment (joukevandermaas) [inline Saule/Serialization/ResourceGraph.cs]: I was half-referring to #99. If I have an object graph like this, where every property is lazy (say, using an ORM):

```
- Person
  - Job
    - LiterallyEveryEmailEverSent     // huge!
```

but no Saule resource ever refers to `LiterallyEveryEmailEverSent`:

```csharp
class PersonResource : ApiResource
{
  public PersonResource() {
    BelongsTo<CompanyResource>(nameof(Person.Job));
  }
}

class CompanyResource : ApiResource
{
  public CompanyResource() {
  }
}
```

The current serializer will materialize `LiterallyEveryEmailEverSent`, which is obviously wasteful because it won't end up in the final JSON output. Does the new serializer still have this effect, or is it circumvented somehow? It's not immediately obvious to me from the code (these things can be tricky).

Since the current serializer *does* materialize everything, I think it's OK if the new one does too; we can always improve on it later. That said, if it already *doesn't* do it, that would be better :smile: In that case, I think tests are necessary, because it seems pretty easy to mess up somewhere and evaluate everything when someone changes something later.

Comment (laurence79) [inline Saule/Serialization/ResourceGraph.cs]: Ok understood. The new serialiser only reads object properties that are declared as attributes or relationships in the resource. So in this case, `LiterallyEveryEmailEverSent` will not be read. I've tidied up the code slightly to make this intent clearer.

I've added a test case to cover this, and also added functionality to serialise dictionaries and dynamics. The old serialiser would have handled this, but originally there were no tests for this specifically.

**Comment by laurence79** [comment]
Ok @joukevandermaas, I think this is ready to be reviewed again 😊

**Comment by joukevandermaas** [comment]
@laurence79 Thanks a lot for this! I published a new pre-release with these changes in it.
